library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/RecipeLoadingEvents.java

```diff
@@ -0,0 +1,285 @@
+		 */
+		@ApiStatus.NonExtendable
+		interface RecipeHandler {
+			/**
+			 * Registers a recipe into the {@link net.minecraft.recipe.RecipeManager}.
+			 * <p>
+			 * The recipe factory is only called if the recipe can be registered.
+			 *
+			 * @param id      identifier of the recipe
+			 * @param factory the recipe factory
+			 */
+			void register(Identifier id, Function<Identifier, Recipe<?>> factory);
+		}
+	}
+
+	/**
+	 * Callback called to modify or replace recipes after recipes are loaded.
+	 */
+	@FunctionalInterface
+	public interface RecipeModifyCallback extends EventAwareListener {
+		/**
+		 * Called after recipes are loaded to modify and replace recipes.
+		 *
+		 * @param handler the recipe handler
+		 */
+		void onModifyRecipe(RecipeHandler handler);
+
+		/**
+		 * This interface should not be extended by users.
+		 */
+		@ApiStatus.NonExtendable
+		interface RecipeHandler {
+			/**
+			 * Replaces a recipe in the {@link net.minecraft.recipe.RecipeManager}.
+			 *
+			 * @param recipe the recipe
+			 */
+			void replace(Recipe<?> recipe);
+
+			/**
+			 * Returns the recipe type of the specified recipe.
+			 *
+			 * @param id the identifier of the recipe
+			 * @return the recipe type if the recipe is present, else {@code null}
+			 */
+			@Nullable RecipeType<?> getTypeOf(Identifier id);
+
+			/**
+			 * Returns whether the {@link net.minecraft.recipe.RecipeManager} contains the specified recipe.
+			 *
+			 * @param id the identifier of the recipe
+			 * @return {@code true} if the recipe is present in the {@link net.minecraft.recipe.RecipeManager}, else {@code false}
+			 */
+			boolean contains(Identifier id);
+
+			/**
+			 * Returns whether the {@link net.minecraft.recipe.RecipeManager} contains the specified recipe of the specified recipe type.
+			 *
+			 * @param id   the identifier of the recipe
+			 * @param type the type of the recipe
+			 * @return {@code true} if the recipe is present in the {@link net.minecraft.recipe.RecipeManager}, else {@code false}
+			 */
+			boolean contains(Identifier id, RecipeType<?> type);
+
+			/**
+			 * Returns the recipe in {@link net.minecraft.recipe.RecipeManager} from its identifier.
+			 *
+			 * @param id the identifier of the recipe
+			 * @return the recipe if present, else {@code null}
+			 */
+			@Nullable Recipe<?> getRecipe(Identifier id);
+
+			/**
+			 * Returns the recipe of the specified recipe type in {@link net.minecraft.recipe.RecipeManager} from its identifier.
+			 *
+			 * @param id   the identifier of the recipe
+			 * @param type the type of the recipe
+			 * @param <T>  the type of the recipe
+			 * @return the recipe if present and of the correct type, else {@code null}
+			 */
+			@Nullable <T extends Recipe<?>> T getRecipe(Identifier id, RecipeType<T> type);
+
+			/**
+			 * Returns all registered recipes.
+			 *
+			 * @return the registered recipes
+			 */
+			Map<RecipeType<?>, Map<Identifier, Recipe<?>>> getRecipes();
```

Comment:
Turn this into a `Table<RecipeType<?>, Identifier, Recipe<?>>`, maybe?

Reply:
Not sure this is a good idea, this is exactly the same representation as Mojang is using in `RecipeManager`.

Reply:
why doesn't Mojang use tables then >:V